libc-top-half/musl/src/process/execve.c

```diff
@@ -1,19 +1,44 @@
 #include <unistd.h>
+#include <string.h>
+#include <stdio.h>
+
 #ifdef __wasilibc_unmodified_upstream
 #include "syscall.h"
 #else
 #include <stdlib.h>
+#include <wasi/api.h>
+#endif
+
+#if !defined(__wasilibc_unmodified_upstream)
+static char* combine_strings(char* const strings[]) {
+	int combined_len = 0;
+	for (char **ptr = (char **)strings; *ptr != NULL; ptr++) {
+		combined_len += strlen(*ptr) + 1;
+	}
+
+	char *combined = malloc((combined_len + 1));
+	char *combined_p = combined;
+	for (char **ptr = (char **)strings; *ptr != NULL; ptr++) {
+		memcpy(combined_p, *ptr, strlen(*ptr));
+		combined_p += strlen(*ptr);
+		*combined_p = '\n';
```

Comment:
I'm not sure about this code, nor separator. Are environment variables separated by a `\n`?
Same as argv. I'm not sure about this impl.

cc @theduke @john-sharratt

Reply:
I'm not opposed to changing the formatting, but I'd like to mention that this isn't new code. This is merely refactored to be reusable for both `argv` and `envp`. This was always used for `argv`.

Reply:
Env vars are separated by `=` when you call into the runtime to get them. This is the other way around. When we want to pass the `envp` to the runtime we combine it into a single string using a separator. I think `\n` is a good token since it cannot be present in key values of an env var.

Reply:
So basically this:
```
["foo=bar", "bar=baz"]
```
will become:
```
"foo=bar\nbar=baz"
```
This will be decoded in the runtime in `proc_exec2`.